bug(codegen): docs generator fails when nested module has same name as function or resource

### What happened?

As described in https://github.com/pulumi/pulumi/pull/14308#issuecomment-1774812922 and https://github.com/pulumi/pulumi/pull/14308#issuecomment-1777972676 the docs generator in `codegen/docs/gen.go` has bug that leads to a panic when a module has the same name as a resource or function within it. 

The `moduleConflictResolver` should prevnt this to happen but it does not.

```go
panic: fatal: An assertion has failed: duplicate file: router/bgp/_index.md

goroutine 1 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/vagrant/code/pulumi/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assertf(0x30?, {0x1ae9498?, 0xc00d629530?}, {0xc000bcf040?, 0x48797e?, 0x1a24940?})
        /home/vagrant/code/pulumi/sdk/go/common/util/contract/assert.go:35 +0xed
github.com/pulumi/pulumi/pkg/v3/codegen.Fs.Add(...)
        /home/vagrant/code/pulumi/pkg/codegen/utilities.go:180
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*modContext).gen.func1({0x0, 0x0}, {0x1adb5af, 0xa}, {0x1a24940, 0xc00e3303c0})
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:1854 +0x225
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*modContext).gen(0xc000bf2aa0, 0xc000bede00)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:1962 +0x11db
github.com/pulumi/pulumi/pkg/v3/codegen/docs.(*docGenContext).generatePackage(0xc0000c1400, {0x0?, 0x0?}, 0x0?)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:2201 +0x2e8
github.com/pulumi/pulumi/pkg/v3/codegen/docs.GeneratePackage(...)
        /home/vagrant/code/pulumi/pkg/codegen/docs/gen.go:2255
github.com/pulumi/registrygen/pkg.generateDocsFromSchema({0x7ffeb02ae50a, 0xf}, 0xc4a256?)
        /home/vagrant/code/pulumi-registrygen/pkg/docs.go:218 +0x54
github.com/pulumi/registrygen/pkg.GenerateDocs({0x7ffeb02ae493?, 0x0?}, {0x7ffeb02ae4b8, 0x6}, {0x7ffeb02ae4cc, 0x30}, {0x7ffeb02ae50a, 0xf}, {0x7ffeb02ae532, 0xb}, ...)
        /home/vagrant/code/pulumi-registrygen/pkg/docs.go:111 +0x896
github.com/pulumi/registrygen/cmd/docs.PackageDocsCmd.func1(0xc0007ff800?, {0x1ad0743?, 0x9?, 0x9?})
        /home/vagrant/code/pulumi-registrygen/cmd/docs/cli.go:89 +0x74
github.com/spf13/cobra.(*Command).execute(0xc0007ff800, {0xc0006943f0, 0x9, 0x9})
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc0007fe600)
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(0xc000124120?)
        /home/vagrant/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992 +0x19
main.main()
        /home/vagrant/code/pulumi-registrygen/main.go:30 +0x8b
```

### Example

**Example:**

Module: `Router` Resource: `Bgp`: path for `resource.tmpl` file = `router/bgp/_index.md`
Module: `Router/Bgp`: path for module _index.md file = `router/bgp/_index.md`

### Output of `pulumi about`

The `registrygen` tool is build using 

```
github.com/pulumi/pulumi/pkg/v3 v3.95.0
github.com/pulumi/pulumi/sdk/v3 v3.95.0
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20240104--docs--fixes-docs-generator-parent-module-computation.yaml
+++ b/changelog/pending/20240104--docs--fixes-docs-generator-parent-module-computation.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: docs
+  description: Fixes docs generator parent module computation

--- a/pkg/codegen/docs/gen.go
+++ b/pkg/codegen/docs/gen.go
@@ -44,6 +44,7 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/codegen/python"
 	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/slice"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 )
 
@@ -2016,25 +2017,30 @@ func (dctx *docGenContext) getMod(
 	add bool,
 ) *modContext {
 	modName := pkg.TokenToModule(token)
-	mod, ok := modules[modName]
+	return dctx.getModByModName(pkg, tokens.ModuleName(modName), tokenPkg, modules, tool, add)
+}
+
+func (dctx *docGenContext) getModByModName(
+	pkg schema.PackageReference,
+	modName tokens.ModuleName,
+	tokenPkg schema.PackageReference,
+	modules map[string]*modContext,
+	tool string,
+	add bool,
+) *modContext {
+	mod, ok := modules[string(modName)]
 	if !ok {
 		mod = &modContext{
 			pkg:           pkg,
-			mod:           modName,
+			mod:           string(modName),
 			tool:          tool,
 			docGenContext: dctx,
 		}
 
 		if modName != "" && codegen.PkgEquals(tokenPkg, pkg) {
-			parentName := path.Dir(modName)
-			// If the parent name is blank, it means this is the package-level.
-			if parentName == "." || parentName == "" {
-				parentName = ":index:"
-			} else {
-				parentName = ":" + parentName + ":"
-			}
-			parent := dctx.getMod(pkg, parentName, tokenPkg, modules, tool, add)
-			if add {
+			parentName, hasParent := parentModule(modName)
+			if add && hasParent {
+				parent := dctx.getModByModName(pkg, parentName, tokenPkg, modules, tool, add)
 				parent.children = append(parent.children, mod)
 			}
 		}
@@ -2042,7 +2048,7 @@ func (dctx *docGenContext) getMod(
 		// Save the module only if we're adding and it's for the current package.
 		// This way, modules for external packages are not saved.
 		if add && tokenPkg == pkg {
-			modules[modName] = mod
+			modules[string(modName)] = mod
 		}
 	}
 	return mod
@@ -2206,6 +2212,11 @@ func (dctx *docGenContext) generatePackage(tool string, pkg *schema.Package) (ma
 	return files, nil
 }
 
+const (
+	// "" indicates the top-most module.
+	topMostModule tokens.ModuleName = ""
+)
+
 // GeneratePackageTree returns a navigable structure starting from the top-most module.
 func (dctx *docGenContext) generatePackageTree() ([]PackageTreeItem, error) {
 	if dctx.modules() == nil {
@@ -2215,8 +2226,7 @@ func (dctx *docGenContext) generatePackageTree() ([]PackageTreeItem, error) {
 	defer glog.Flush()
 
 	var packageTree []PackageTreeItem
-	// "" indicates the top-most module.
-	if rootMod, ok := dctx.modules()[""]; ok {
+	if rootMod, ok := dctx.modules()[string(topMostModule)]; ok {
 		tree, err := generatePackageTree(*rootMod)
 		if err != nil {
 			glog.Errorf("Error generating the package tree for package: %v", err)
@@ -2258,4 +2268,17 @@ func GeneratePackage(tool string, pkg *schema.Package) (map[string][]byte, error
 // GeneratePackageTree returns a navigable structure starting from the top-most module.
 func GeneratePackageTree() ([]PackageTreeItem, error) {
 	return defaultContext.generatePackageTree()
+}
+
+// Returns the parent module, if available. For top-level modules the parent is a special
+// topMostModule. For topMostModule itself there is no parent and this function returns false.
+func parentModule(modName tokens.ModuleName) (tokens.ModuleName, bool) {
+	switch {
+	case modName == topMostModule:
+		return "", false
+	case strings.Contains(string(modName), tokens.QNameDelimiter):
+		return tokens.ModuleName(tokens.QName(modName).Namespace()), true
+	default:
+		return topMostModule, true
+	}
 }

--- a/pkg/codegen/docs/package_tree_test.go
+++ b/pkg/codegen/docs/package_tree_test.go
@@ -87,9 +87,6 @@ func TestGeneratePackageTreeNested(t *testing.T) {
 			schema.PackageSpec{
 				Name:    "fortios",
 				Version: "0.0.1",
-				Meta: &schema.MetadataSpec{
-					ModuleFormat: "(.*)(?:/[^/]*)",
-				},
 				Resources: map[string]schema.ResourceSpec{
 					"fortios:router/bgp:Bgp":             {},
 					"fortios:router/bgp/network:Network": {},
@@ -133,9 +130,6 @@ func TestGeneratePackageTreeNested(t *testing.T) {
 			schema.PackageSpec{
 				Name:    "fortios",
 				Version: "0.0.1",
-				Meta: &schema.MetadataSpec{
-					ModuleFormat: "(.*)(?:/[^/]*)",
-				},
 				Resources: map[string]schema.ResourceSpec{
 					"fortios:log/syslogd/v2/filter:Filter":                   {},
 					"fortios:log/syslogd/v2/overridefilter:Overridefilter":   {},
@@ -210,33 +204,44 @@ func TestGeneratePackageTreeNested(t *testing.T) {
 	for _, tc := range testCases {
 		tc := tc
 
-		t.Run(tc.name, func(t *testing.T) {
-			t.Parallel()
-
-			c, _ := prep(t, tc)
-
-			items, err := c.generatePackageTree()
-			require.NoError(t, err)
-
-			data, err := json.MarshalIndent(items, "", "  ")
-			require.NoError(t, err)
-
-			t.Logf("%s", string(data))
-
-			require.JSONEq(t, tc.expect, string(data))
-		})
-
-		t.Run(tc.name+"/generatePackage", func(t *testing.T) {
-			t.Parallel()
-
-			c, schemaPkg := prep(t, tc)
-
-			files, err := c.generatePackage("test", schemaPkg)
-			require.NoError(t, err)
-
-			for f := range files {
-				t.Logf("+ %v", f)
+		for _, mv := range []string{"", "(.*)(?:/[^/]*)"} {
+			name := tc.name
+			if mv != "" {
+				tc.spec.Meta = &schema.MetadataSpec{
+					ModuleFormat: mv,
+				}
+				name += "-withModuleFormat"
+			} else {
+				name += "-withoutModuleFormat"
 			}
-		})
+			t.Run(name, func(t *testing.T) {
+				t.Parallel()
+
+				c, _ := prep(t, tc)
+
+				items, err := c.generatePackageTree()
+				require.NoError(t, err)
+
+				data, err := json.MarshalIndent(items, "", "  ")
+				require.NoError(t, err)
+
+				t.Logf("%s", string(data))
+
+				require.JSONEq(t, tc.expect, string(data))
+			})
+
+			t.Run(name+"/generatePackage", func(t *testing.T) {
+				t.Parallel()
+
+				c, schemaPkg := prep(t, tc)
+
+				files, err := c.generatePackage("test", schemaPkg)
+				require.NoError(t, err)
+
+				for f := range files {
+					t.Logf("+ %v", f)
+				}
+			})
+		}
 	}
 }

--- a/pkg/codegen/docs/package_tree_test.go
+++ b/pkg/codegen/docs/package_tree_test.go
@@ -15,9 +15,11 @@
 package docs
 
 import (
+	"encoding/json"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
 	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
 )
@@ -67,4 +69,174 @@ func TestGeneratePackageTree(t *testing.T) {
 		assert.Equal(t, entryTypeResource, children[0].Type)
 		assert.Equal(t, entryTypeFunction, children[1].Type)
 	})
+}
+
+// Original issues: pulumi/pulumi#14821, pulumi/pulumi#14820.
+func TestGeneratePackageTreeNested(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		name   string
+		spec   schema.PackageSpec
+		expect string
+	}
+
+	testCases := []testCase{
+		{
+			"14820",
+			schema.PackageSpec{
+				Name:    "fortios",
+				Version: "0.0.1",
+				Meta: &schema.MetadataSpec{
+					ModuleFormat: "(.*)(?:/[^/]*)",
+				},
+				Resources: map[string]schema.ResourceSpec{
+					"fortios:router/bgp:Bgp":             {},
+					"fortios:router/bgp/network:Network": {},
+				},
+			},
+			`
+			[
+			  {
+			    "name": "router",
+			    "type": "module",
+			    "link": "router/",
+			    "children": [
+			      {
+				"name": "bgp",
+				"type": "module",
+				"link": "bgp/",
+				"children": [
+				  {
+				    "name": "Network",
+				    "type": "resource",
+				    "link": "network"
+				  }
+				]
+			      },
+			      {
+				"name": "Bgp",
+				"type": "resource",
+				"link": "bgp"
+			      }
+			    ]
+			  },
+			  {
+			    "name": "Provider",
+			    "type": "resource",
+			    "link": "provider"
+			  }
+			]`,
+		},
+		{
+			"14821",
+			schema.PackageSpec{
+				Name:    "fortios",
+				Version: "0.0.1",
+				Meta: &schema.MetadataSpec{
+					ModuleFormat: "(.*)(?:/[^/]*)",
+				},
+				Resources: map[string]schema.ResourceSpec{
+					"fortios:log/syslogd/v2/filter:Filter":                   {},
+					"fortios:log/syslogd/v2/overridefilter:Overridefilter":   {},
+					"fortios:log/syslogd/v2/overridesetting:Overridesetting": {},
+					"fortios:log/syslogd/v2/setting:Setting":                 {},
+				},
+			},
+			`
+			[
+			  {
+			    "name": "log",
+			    "type": "module",
+			    "link": "log/",
+			    "children": [
+			      {
+				"name": "syslogd",
+				"type": "module",
+				"link": "syslogd/",
+				"children": [
+				  {
+				    "name": "v2",
+				    "type": "module",
+				    "link": "v2/",
+				    "children": [
+				      {
+					"name": "Filter",
+					"type": "resource",
+					"link": "filter"
+				      },
+				      {
+					"name": "Overridefilter",
+					"type": "resource",
+					"link": "overridefilter"
+				      },
+				      {
+					"name": "Overridesetting",
+					"type": "resource",
+					"link": "overridesetting"
+				      },
+				      {
+					"name": "Setting",
+					"type": "resource",
+					"link": "setting"
+				      }
+				    ]
+				  }
+				]
+			      }
+			    ]
+			  },
+			  {
+			    "name": "Provider",
+			    "type": "resource",
+			    "link": "provider"
+			  }
+			]`,
+		},
+	}
+
+	prep := func(t *testing.T, tc testCase) (*docGenContext, *schema.Package) {
+		t.Helper()
+
+		schemaPkg, err := schema.ImportSpec(tc.spec, nil)
+		assert.NoError(t, err, "importing spec")
+
+		c := newDocGenContext()
+		c.initialize("test", schemaPkg)
+
+		return c, schemaPkg
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			c, _ := prep(t, tc)
+
+			items, err := c.generatePackageTree()
+			require.NoError(t, err)
+
+			data, err := json.MarshalIndent(items, "", "  ")
+			require.NoError(t, err)
+
+			t.Logf("%s", string(data))
+
+			require.JSONEq(t, tc.expect, string(data))
+		})
+
+		t.Run(tc.name+"/generatePackage", func(t *testing.T) {
+			t.Parallel()
+
+			c, schemaPkg := prep(t, tc)
+
+			files, err := c.generatePackage("test", schemaPkg)
+			require.NoError(t, err)
+
+			for f := range files {
+				t.Logf("+ %v", f)
+			}
+		})
+	}
 }

--- a/pkg/codegen/docs/package_tree_test.go
+++ b/pkg/codegen/docs/package_tree_test.go
@@ -82,6 +82,58 @@ func TestGeneratePackageTreeNested(t *testing.T) {
 	}
 
 	testCases := []testCase{
+
+		{
+			"index top module",
+			schema.PackageSpec{
+				Name:    "testindex",
+				Version: "0.0.1",
+				Resources: map[string]schema.ResourceSpec{
+					"testindex:index:Resource": {},
+				},
+			},
+			`[
+			 {
+			   "name": "Provider",
+			   "type": "resource",
+			   "link": "provider"
+			 },
+			 {
+			   "name": "Resource",
+			   "type": "resource",
+			   "link": "resource"
+			 }
+		       ]`,
+		},
+		{
+			"random",
+			schema.PackageSpec{
+				Name:    "random",
+				Version: "0.0.1",
+				Resources: map[string]schema.ResourceSpec{
+					"random:index/randomId:RandomId":             {},
+					"random:index/randomPassword:RandomPassword": {},
+				},
+			},
+			`
+			 [
+			  {
+			    "name": "Provider",
+			    "type": "resource",
+			    "link": "provider"
+			  },
+			  {
+			    "name": "RandomId",
+			    "type": "resource",
+			    "link": "randomid"
+			  },
+			  {
+			    "name": "RandomPassword",
+			    "type": "resource",
+			    "link": "randompassword"
+			  }
+			]`,
+		},
 		{
 			"14820",
 			schema.PackageSpec{

--- a/pkg/codegen/docs/package_tree_test.go
+++ b/pkg/codegen/docs/package_tree_test.go
@@ -82,7 +82,6 @@ func TestGeneratePackageTreeNested(t *testing.T) {
 	}
 
 	testCases := []testCase{
-
 		{
 			"index top module",
 			schema.PackageSpec{

